### PR TITLE
Implement `.satisfies()` and `.maxSatisfying()` methods

### PR DESCRIPTION
**Pull request by LucianBuzzo**
Connects to #23

change-type: minor

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -282,3 +282,59 @@ export const gt = (versionA: string | null, versionB: string | null): boolean =>
...
+
+	if (!semver.valid(version)) {
+		return false;
+	}
+
+	return semver.satisfies(version, range);
+};
+
+/**
+ * @summary Return the lowest version in the list that satisfies the range
```
Comment (pimterry) [inline src/index.ts]: This should be 'highest'

File: src/index.ts [hunk 2]
```diff
@@ -282,3 +282,59 @@ export const gt = (versionA: string | null, versionB: string | null): boolean =>
...
+ * @name maxSatisfying
+ * @public
+ * @function
+ *
+ * @description Return the highest version in the list that satisfies the range, or null if none of them do.
```
Comment (pimterry) [inline src/index.ts]: As above, this should be explicit that non-semver never satisfies anybody

Comment (pimterry) [inline src/index.ts]: It would also be good to point out the semantics for equally highest values: if there are multiple equally high values, the first one found in the array is returned.

File: test/index.spec.ts [hunk 3]
```diff
@@ -304,4 +304,36 @@ describe('resin-semver', () => {
...
+		});
+
+		it('should correctly evaluate invalid semver values', () => {
+			expect(semver.satisfies('Linux 14.04', testRange)).to.equal(false);
+			expect(semver.satisfies('A development version', testRange)).to.equal(false);
+			expect(semver.satisfies('Version A', testRange)).to.equal(false);
+		});
+
+		it('should correctly evaluate null values', () => {
+			expect(semver.satisfies(null, testRange)).to.equal(false);
```
Comment (pimterry) [inline test/index.spec.ts]: For the null/non-semver cases, I'd also include a quick couple of surprising cases (`*`, `< 1.0.0`), as both documentation and protection against accidental changes in this behaviour later, since that seems especially likely with debatable semantics like these.

File: src/index.ts [hunk 4]
```diff
@@ -282,3 +282,59 @@ export const gt = (versionA: string | null, versionB: string | null): boolean =>
...
 	return compare(versionA, versionB) < 0;
 };
+
+/**
+ * @summary Check if a version satisfies a range
+ * @name satisfies
+ * @public
+ * @function
+ *
+ * @description Return true if the version satisfies the range
```
Comment (thgreasi) [inline src/index.ts]: So it could be something like `Returns true when the parsed version satisfies the range`?

File: src/index.ts [hunk 5]
```diff
@@ -282,3 +282,59 @@ export const gt = (versionA: string | null, versionB: string | null): boolean =>
...
+ * docs for details
+ *
+ * @returns {string|null} - The highest matching version string, or null.
+ *
+ */
+export const maxSatisfying = (versions: Array<string | null>, range: string) => {
+	let max: string | null = null;
+
+	versions.forEach((version) => {
+		if (satisfies(version, range) && gt(version, max)) {
```
Comment (thgreasi) [inline src/index.ts]: How about `max === null || gt(version, max)` to save some nested calls?

My only heads-up is that the `satisfies` in here will run `normalize` & `semver.valid` N times.
How about normalizing & checking for validity outside of the loop and then use the `semver.satisfies`.

Getting with it even further, someone might think that `maxSatisfying` can be the base plural/composite implementation of `satisfies`.
In other words:
```js
satisfies(version, range) = maxSatisfying([version], range) != null;
```
@pimterry ?

Comment (pimterry) [inline src/index.ts]: > My only heads-up is that the satisfies in here will run normalize & semver.valid N times.

Personally I'm not super worried by this - we're unlikely to be checking maxSatisfying in any kind of performance critical situations (the download modal taking an extra 1ms to load is totally fine), and I think it'll make the overall code a meaningful amount worse: it's nice to have this directly depend on `satisfies`, because that guarantees that we get the same semantics. Doesn't make a big difference, and if people have strong opinions let's go for it, but by default I'd prefer nice code to microoptimized code, if it's not critical.

> Getting with it even further, someone might think that maxSatisfying can be the base plural/composite implementation of satisfies. In other words:
> ```
> satisfies(version, range) = maxSatisfying([version], range) != null;
> ```

Ah, interesting, you're saying that we could implement this like that? That would solve some of my concerns, by tying the two together. Sounds like it might make the overall code harder to follow, but I'm not sure - up to you @LucianBuzzo

Comment (thgreasi) [inline src/index.ts]: Just an idea. :+1: on the already simple & clear implementation.

Comment (LucianBuzzo) [inline src/index.ts]: I'm happy with the implementation as it stands, it's clear and easy IMO
